Make Scala compilation 26% faster

jadenPete opened the pull request:
This PR speeds up Scala compilation with this ruleset by ~26% across-the-board and substantially reduces the complexity of the ruleset by using the compiler bridge directly instead of Zinc. Note that all the goodies we get from Zinc and its analysis information are **not** exclusive to Zinc and are instead gotten by using the compiler bridge, which Zinc invokes under the hood. However, this PR does drop support for incremental compilation.

Please read the commit messages (especially the second commit description) for more information on the technical context surrounding this decision.

Commit:
Cleaned up the imports in src/main/scala

Commit:
Sidestep Zinc and use the compiler interface directly

Basic benchmarking shows that we're incurring a ~49% overhead from using
Zinc instead of the Scala compiler directly. That is, we wrote a basic
worker that invokes the Scala 2 compiler programmatically via
`scala.tools.nsc.Main`, and found that it was ~49% faster
across-the-board.

This doesn't necessarily mean that we can speed up `ZincRunner` by 49%.
A lot of that time is spent in compilation phases
[added](https://github.com/scala/scala/blob/2.13.x/src/sbt-bridge/scala/tools/xsbt/CallbackGlobal.scala#L166)
by the compiler bridge (implementation of the compiler interface)
itself. There are three of these phases; they're responsible for:
1. Mapping generated `.class` files to sources
2. Analyzing depedendencies between code
3. Producing the analysis information outputted by Zinc

We rely on 2 and 3 for:
- Dependency checking
- Detecting main classes
- Test discovery

We only use 2 and 3, and although it's likely the phases could be
slimmed down to only extract the information we need, implementing them
ourselves would add a lot of complexity to `ZincRunner` and make it
less portable . There's also no guarantee
that we could do a better job.

What can be eliminated is all the overhead related to incremental
compilation. We haven't used incremental compilation for years and have
no intention of re-enabling it.

This commit isn't about ripping out Zinc entirely; we still use Zinc for
the following things:
- Classloading the compiler interface from the compiler classpath
- Implementations of various `xsbti.*` traits and interfaces

Commit:
Do test discovery in ZincRunner

Commit:
Ripped out the remaining Zinc-related code

Commit:
Updated the documentation

jnowjack-lucidchart commented on src/main/scala/sbt/internal/inc/classfile/package.scala:
```diff
@@ -0,0 +1,59 @@
+    val classloader = ClasspathUtil.toLoader(outputDirectory +: classpath)
+
+    /**
+     * TODO: Make this more similar to the `readAPI` defined in [[sbt.internal.inc.javac.AnalyzingJavaCompiler]].
```
Is this TODO a blocker for our use cases?

jadenPete replied:
I don't think so. If it were, we'd see problems with test discovery in mixed-compilation targets (i.e. targets with Scala and Java code), but we don't have any of those in our monorepo because all of our tests are written in Scala. I added this comment because I think it's something we should figure out before we bring our workers into the mainline ruleset.

jnowjack-lucidchart commented on src/main/scala/sbt/internal/inc/classfile/package.scala:
Do you think we could ask to see if the functions you're using could be moved to a Util class instead of private in an internal package?

jadenPete replied:
We could, but I don't think the effort would be worthwhile, especially considering that the Scala maintainers have expressed in the past that Zinc/compiler bridge stuff isn't meant to be a public interface.

If we're ever using the compiler bridge in the mainline ruleset, I think it'd be good to start up conversations with the Scala maintainers about designing a build system-agnostic, public interface to the Scala compilers. If such an interface existed, this file wouldn't need to exist.

jnowjack-lucidchart commented on src/main/scala/higherkindness/rules_scala/common/sbt-testing/fingerprints.scala:
```diff
@@ -1,23 +1,38 @@
+sealed trait TestFingerprint extends Fingerprint
 
-class TestAnnotatedFingerprint(val annotationName: String, val isModule: Boolean)
+object TestFingerprint {
```
Is there anything in sbt.testing.Fingerpint that could be nondeterministic?

jadenPete replied:
Potentially, but we never serialize any of the fields in `Fingerprint`.

jnowjack-lucidchart commented on src/main/scala/higherkindness/rules_scala/common/sbt-testing/fingerprints.scala:
```diff
@@ -1,23 +1,38 @@
 }
 
-class TestSubclassFingerprint(val isModule: Boolean, val requireNoArgConstructor: Boolean, val superclassName: String)
+case class TestSubclassFingerprint(isModule: Boolean, requireNoArgConstructor: Boolean, val superclassName: String)
```
```suggestion
case class TestSubclassFingerprint(isModule: Boolean, requireNoArgConstructor: Boolean, superclassName: String)
```

jadenPete replied:
Good catch!

jnowjack-lucidchart commented on src/main/scala/higherkindness/rules_scala/workers/zinc/test/TestRunner.scala:
```diff
@@ -192,8 +191,10 @@ object TestRunner {
-    val loader = new TestFrameworkLoader(classLoader, logger)
-    val frameworks = testRunnerRequest.frameworks.flatMap(loader.load)
+    val loader = new TestFrameworkLoader(classLoader)
+    val testsFile = Using(new FileInputStream(testRunnerRequest.testsFile.toString)) { stream =>
```
minor nit: `testsFileData` or something might be nice

jadenPete replied:
Good point. `TestsFile` is a bit of a misnomer. I'll rename it to `TestsFileData`, and `testsFile` to `testsFileData`.

jnowjack-lucidchart commented on tests/test-frameworks/mixed/BUILD:
```diff
@@ -3,15 +3,17 @@ load("@rules_scala_annex//rules:scala.bzl", "scala_test")
 scala_test(
     name = "mixed_2_13",
     srcs = glob(["*.scala"]),
+    deps_used_whitelist = [
```
Could you add a comment explaining this change? Might be worth a TODO to see if we can get the deps used to not mark these things required for test discovery as unused.

coreywoodfield replied:
Are these not runtime deps? :thinking: If we don't depend on them but we need them it seems like they're runtime deps

jadenPete replied:
They aren't, no. These tests aren't needed at runtime; they're needed for test discovery. The reason they have to be in `deps` now is because test discovery now happens at compile-time, not at runtime. I haven't updated the deps checker to mark these as used because in most cases, the target containing the implementation of the SBT test interface is the same one that the target would've depended on anyway. That's why none of the other external dependencies are in this list.

I can fix this, though. One important thing to note is that this will break a property of our dependency checker that's been thus-far guaranteed: that it only allows a dependency to be in `deps` or not in `deps`. When we do test discovery, we go down a big list of test "frameworks". If we can find the test framework in the classpath, we look for tests that use that framework; otherwise, we don't. This means that, for example, if specs2 isn't on the classpath, but the target contains specs2 tests, the tests will silently fail to execute.

I think this is fine and unavoidable, but you disagree, I think we should leave this as-is.

coreywoodfield commented on docs/stardoc/scala_proto.md:
```diff
@@ -73,7 +73,8 @@ toolchain(
 | :------------- | :------------- | :------------- | :------------- | :------------- |
 | <a id="scala_proto_toolchain-name"></a>name |  A unique name for this target.   | <a href="https://bazel.build/concepts/labels#target-names">Name</a> | required |  |
-| <a id="scala_proto_toolchain-compiler"></a>compiler |  The compiler to use to generate Scala form proto sources   | <a href="https://bazel.build/concepts/labels">Label</a> | optional |  `None`  |
+| <a id="scala_proto_toolchain-compiler"></a>compiler |  The compiler to use to generate Scala form proto sources   | <a href="https://bazel.build/concepts/labels">Label</a> | required |  |
```
```suggestion
| <a id="scala_proto_toolchain-compiler"></a>compiler |  The compiler to use to generate Scala from proto sources   | <a href="https://bazel.build/concepts/labels">Label</a> | required |  |
```

jadenPete replied:
This documentation is generated, but I'll update the string that contains this word.

coreywoodfield commented on MODULE.bazel:
```diff
@@ -41,6 +41,7 @@ annex.install(
         "com.thesamet.scalapb:scalapb-runtime_2.13:{}".format(scalapb_version),
         "net.sourceforge.argparse4j:argparse4j:0.8.1",
         "org.jacoco:org.jacoco.core:0.7.5.201505241946",
+        "org.playframework:play-json_3:3.0.4",
```
play json is pretty big... (see 224 new lines in `annex_install.json`). Is there a more lightweight approach that would be almost as good?

jadenPete replied:
Potentially; most of the added dependencies are Jackson dependencies, so we probably don't want to use Jackson.

[uPickle](https://github.com/com-lihaoyi/upickle) looks promising; it's [recommended](https://docs.scala-lang.org/toolkit/json-intro.html) in the official Scala documentation and is written by the same guy who wrote Ammonite and Mill.

I'll try to swap out Play JSON for it.

jadenPete replied:
Update: It ended up being similar in number of transitive dependencies (6 instead of 7), so I've decided to keep Play JSON.

Commit:
Consider targets used for test discovery in dependency checking

Before, we did test discovery at runtime, but we now do it at
compile-time, for speed and to avoid having to serialize the analysis
information returned by the compiler bridge.

This commits marks targets used during test discovery as used, thereby
subjecting them to depednency checking. In most cases, this won't make a
difference as the target containing the implementation of the
SBT test interface is the same one that the target being built would've
depended on anyway.

One important thing to note is that this will break a property of our
dependency checker that's been thus-far guaranteed: that it only allows
a dependency to be in `deps` or not in `deps`. When we do
test discovery, we iterate through a list of test "frameworks". If we
find a test framework in the classpath, we look for tests that use that
framework; otherwise, we don't. This means that, for example, if specs2
isn't on the classpath, but the target contains specs2 tests, the tests
will silently fail to execute.

This means that now, having that a dependency in `deps` or not in `deps`
are sometimes both valid states to be in. Prior to this change, that
wasn't the case; there was a single, correct list of dependencies.

Commit:
Corrected a misspelling
